src/lib/fs/path.c

```diff
@@ -293,3 +341,389 @@ make_path_absolute(char *fname)
+                         smartlist_t *result)
+{
+  char *path_until_glob = get_path_until_separator(pattern, start_sep);
+  smartlist_add(result, path_until_glob);
+  // if the following fragments have no globs, we're done
+  const char *pattern_after = &pattern[end_sep+1];
+  return has_glob(pattern_after);
+}
+
+/*****************************************************************************/
+
+#ifdef _WIN32
+/** Returns a list of paths that match <b>pattern</b>. Retruns NULL on error.
+ * Due to Windows API limitations, globs are only accepted on the last path
+ * component. */
+static struct smartlist_t *
+unglob_file_name(const char *pattern)
+{
+  smartlist_t *result;
+  TCHAR tpattern[MAX_PATH] = {0};
+  char name[MAX_PATH*2+1] = {0};
+  HANDLE handle;
+  WIN32_FIND_DATA findData;
+#ifdef UNICODE
+  mbstowcs(tpattern,pattern,MAX_PATH);
+#else
+  strlcpy(tpattern, pattern, MAX_PATH);
+#endif
+  clean_fname_for_stat(tpattern);  // remove trailing backslash
+  if (file_status(tpattern) == FN_DIR) {
+    // special case: we want to return the directory and not the files inside
+    result = smartlist_new();
+    smartlist_add_strdup(result, tpattern);
+    return result;
+  }
+  if (INVALID_HANDLE_VALUE == (handle = FindFirstFile(tpattern, &findData))) {
+    return GetLastError() == ERROR_FILE_NOT_FOUND ? smartlist_new() : NULL;
+  }
+  // remove file name and last path separator from tpattern
+  // used later to create the full path
+  char *lastSep = strrchr(tpattern, *PATH_SEPARATOR);
+  char *lastSepUnix = strrchr(tpattern, '/');
+  lastSep = lastSep == NULL ? tpattern : lastSep;
+  lastSepUnix = lastSepUnix == NULL ? tpattern : lastSepUnix;
+  lastSep = lastSep > lastSepUnix ? lastSep : lastSepUnix;
+  *lastSep = '\0';
+  result = smartlist_new();
+  while (1) {
+#ifdef UNICODE
+    wcstombs(name,findData.cFileName,MAX_PATH);
+    name[sizeof(name)-1] = '\0';
+#else
+    strlcpy(name,findData.cFileName,sizeof(name));
+#endif /* defined(UNICODE) */
+    if (strcmp(name, ".") && strcmp(name, "..")) {
+      char *fullpath;
+      tor_asprintf(&fullpath, "%s"PATH_SEPARATOR"%s", tpattern, name);
+      smartlist_add(result, fullpath);
+    }
+    if (!FindNextFile(handle, &findData)) {
+      DWORD err;
+      if ((err = GetLastError()) != ERROR_NO_MORE_FILES) {
+        char *errstr = format_win32_error(err);
+        log_warn(LD_FS, "Error reading directory '%s': %s", pattern, errstr);
+        tor_free(errstr);
+      }
+      break;
+    }
+  }
+  FindClose(handle);
+  return result;
+}
+#endif /* defined(_WIN32) */
+
+/** If <b>path</b> represents an existing file, add it to <b>results</b>
+ * without the trailing separator. Otherwise do nothing. */
+static void
+add_to_results(const char *path, smartlist_t *results)
+{
+  file_status_t file_type = file_status(path);
+  char *clean_path = tor_strdup(path);
+  clean_fname_for_stat(clean_path);
+  if (file_type != FN_ERROR && file_type != FN_NOENT) {
```

Comment:
This is a good place to use is_file() from files.h

Reply:
Suggestion applied in ae426f9